Remove or replace uses of isinstance()

cjmayo opened the pull request:
More info in the comments.

The one for FtpUrl looks like a fix.

cjmayo commented:
New commit because no need to encode/decode an empty string.

```diff
--- a/linkcheck/strformat.py
+++ b/linkcheck/strformat.py
@@ -67,7 +67,7 @@ def ascii_safe(s):
     @return: version of s containing only ASCII characters, or None if s was None
     @rtype: string or None
     """
-    if s is not None:
+    if s:
         s = s.encode('ascii', 'ignore').decode('ascii')
     return s
```

mgedmin commented on linkcheck/checker/ftpurl.py:
```diff
@@ -115,8 +115,9 @@ def cwd(self):
-            path = path.decode(self.filename_encoding, 'replace')
+        if self.filename_encoding != "utf-8":
+            path = path.encode(self.filename_encoding, 'replace').decode(
+                self.filename_encoding)
```
This change doesn't make sense to me.  Can you drop it from this PR please?

(There might be some logic errors in ftpurl's filename handling, but bugfixes shouldn't hide in PRs that purport to be refactorings.)

cjmayo replied:
Dropped that change.

cjmayo replied:
Raised issue #450.

mgedmin commented on linkcheck/checker/fileurl.py:
```diff
@@ -234,8 +234,7 @@ def read_content(self):
             data = get_index_html(get_files(self.get_os_filename()))
-            if isinstance(data, str):
-                data = data.encode("iso8859-1", "ignore")
+            data = data.encode("iso8859-1", "ignore")
```
This is a correct refactoring, but it brings attention to buggy code.  ISO-8859-1 is not the right choice for arbitrary filenames.  UTF-8 might work better.  Ideally we'd plumb things up so that get_content() that calls get_raw_content() that calls download_content() that calls read_content() wouldn't need to have BeautifulSoup guess what the encoding is, given that it's hardcoded right here!

(Maybe get_index_html() ought to generate a `<meta charset="UTF-8">`.)

In case it wasn't clear: this change here is fine, and the bugfix belongs to a separate PR.

cjmayo replied:
Raised issue #451.

cjmayo commented:
Dropped the "Replace isinstance() in FtpUrl.cwd()" and its revert commit. Renamed the squash commit to "No need to process an empty string in str_format.ascii_safe()".
